src/metabase/lib/join/util.cljc

```diff
@@ -0,0 +1,89 @@
+  "A Join that may not yet have an `:alias`, which is normally required; [[join]] accepts this and will add a default
+  alias if one is not present."
+  [:merge
+   [:ref ::lib.schema.join/join]
+   [:map
+    [:alias {:optional true} [:ref ::lib.schema.join/alias]]]])
+
+(def PartialJoin
+  "A join that may not yet have an `:alias` or `:conditions`."
+  [:merge
+   JoinWithOptionalAlias
+   [:map
+    [:conditions {:optional true} [:ref ::lib.schema.join/conditions]]]])
+
+(def Field
+  "A field in a join, either [[lib.metadata/ColumnMetadata]] or a `:field` ref."
+  [:or
+   lib.metadata/ColumnMetadata
+   [:ref :mbql.clause/field]])
+
+(def FieldOrPartialJoin
+  "A field or a partial join."
+  [:or Field PartialJoin])
+
+(mu/defn current-join-alias :- [:maybe ::lib.schema.common/non-blank-string]
+  "Get the current join alias associated with something, if it has one."
+  [field-or-join :- [:maybe FieldOrPartialJoin]]
+  (case (lib.dispatch/dispatch-value field-or-join)
+    :dispatch-type/nil nil
+    :field             (:join-alias (lib.options/options field-or-join))
+    :metadata/column   (:metabase.lib.join/join-alias field-or-join)
+    :mbql/join         (:alias field-or-join)))
+
+(mu/defn joined-field-desired-alias :- ::lib.schema.common/non-blank-string
+  "Desired alias for a Field that comes from a join, e.g.
+
+    MyJoin__my_field
+
+  You should pass the results thru a unique name function."
+  [join-alias :- ::lib.schema.common/non-blank-string
+   field-name :- ::lib.schema.common/non-blank-string]
+  (lib.util/format "%s__%s" join-alias field-name))
+
+(mu/defn format-implicit-join-name :- ::lib.schema.common/non-blank-string
+  "Name for an implicit join against `table-name` via an FK field, e.g.
+
+    CATEGORIES__via__CATEGORY_ID
+
+  You should make sure this gets ran thru a unique-name fn."
+  [table-name           :- ::lib.schema.common/non-blank-string
+   source-field-id-name :- ::lib.schema.common/non-blank-string]
+  (lib.util/format "%s__via__%s" table-name source-field-id-name))
+
+(defn- implicit-join-name [query {:keys [fk-field-id table-id], :as _field-metadata}]
+  (when (and fk-field-id table-id)
+    (when-let [table (lib.metadata/table-or-card query table-id)]
+      (let [table-name           (:name table)
+            source-field-id-name (:name (lib.metadata/field query fk-field-id))]
+        (format-implicit-join-name table-name source-field-id-name)))))
+
+(mu/defn desired-alias :- ::lib.schema.common/non-blank-string
+  "Desired alias for a Field e.g.
+
+    my_field
+
+    OR
+
+    MyJoin__my_field
+
+  You should pass the results thru a unique name function."
+  [query          :- ::lib.schema/query
+   field-metadata :- lib.metadata/ColumnMetadata]
+  (if-let [join-alias (or (current-join-alias field-metadata)
+                          (implicit-join-name query field-metadata))]
+    (joined-field-desired-alias join-alias (:name field-metadata))
+    (:name field-metadata)))
```

Comment:
I don't really agree with moving this here since even a Field without a join alias still has a desired alias

Reply:
I'd support an alternative home, it just needed out of `metabase.lib.field` to avoid a circular dep.

`metabase.lib.field.util`?